File: sdk/typescript/src/vertex.ts
// VertexClient — TypeScript SDK surface for Weave Vertex (VTX-108).
//
// Mirrors the shape of the existing OmsClient module: a constructor that
// takes a base URL + fetch impl and namespaced calls grouped by resource
// (scenarios, graphs). Scenario runs follow the mounted start-and-poll
// contract: POST /runs accepts a run and GET /runs/{runRid} returns the
// persisted lifecycle record.

export interface VertexClientOptions {
  baseUrl?: string;
  /** Inject a fetch implementation. Defaults to globalThis.fetch. */
  fetch?: typeof fetch;
}

export interface ScenarioCreateInput {
  caseStudyRid: string;
  name: string;
  parentOntologyCommit: string;
}

export interface Scenario {
  rid: string;
  caseStudyRid: string;
  name: string;
  status: 'draft' | 'frozen' | 'archived';
  immutable: boolean;
}

export interface ScenarioRunOptions extends ScenarioRunPollOptions {
  /** Scenario-run streaming is not mounted yet; true rejects without a request. */
  streaming?: boolean;
}

export interface ScenarioRunPollOptions {
  intervalMs?: number;
  timeoutMs?: number;
  signal?: AbortSignal;
}

export interface ScenarioRunStartResponse {
  runRid: string;
  status: 'pending';
}

export interface ScenarioRunCheckpoint {
  runRid: string;
  scenarioRid: string;
  status: ScenarioRunLifecycleStatus;
  completed?: string[];
  lastActivity?: string;
  attemptsById?: Record<string, number>;
  error?: string;
  updatedAt?: string;
}

export interface ScenarioRunRecord {
  rid: string;
  scenarioRid: string;
  status: ScenarioRunLifecycleStatus;
  error?: string;
  checkpoint?: ScenarioRunCheckpoint;
  startedAt?: string;
  completedAt?: string | null;
  createdAt?: string;
}

export type ScenarioRunLifecycleStatus =
  | 'pending'
  | 'running'
  | 'succeeded'
  | 'failed'
  | 'canceled';

export interface ApplyToMainInput {
  scenarioRid: string;
}

export class VertexHttpError extends Error {
  public status: number;
  public path: string;

  constructor(status: number, path: string, body: string) {
    super(`Vertex SDK: ${status} on ${path}: ${body}`);
    this.status = status;
    this.path = path;
  }
}

export class VertexScenarioRunPollingTimeoutError extends Error {
  constructor(timeoutMs: number) {
    super(`Vertex SDK: scenario run polling timed out after ${timeoutMs} ms`);
    this.name = 'VertexScenarioRunPollingTimeoutError';
  }
}

export class VertexScenarioRunPollingAbortedError extends Error {
  constructor() {
    super('Vertex SDK: scenario run polling was aborted');
    this.name = 'VertexScenarioRunPollingAbortedError';
  }
}

export class VertexClient {
  private baseUrl: string;
  private fetchImpl: typeof fetch;

  constructor(opts: VertexClientOptions = {}) {
    this.baseUrl = opts.baseUrl ?? 'http://localhost:9117';
    this.fetchImpl = opts.fetch ?? globalThis.fetch.bind(globalThis);
  }

  get scenarios() {
    return {
      create: (input: ScenarioCreateInput) => this.scenarioCreate(input),
      startRun: (rid: string): Promise<ScenarioRunStartResponse> =>
        this.scenarioStartRun(rid),
      run: (rid: string, opts?: ScenarioRunOptions): Promise<ScenarioRunRecord> =>
        this.scenarioRun(rid, opts),
      getRun: (scenarioRid: string, runRid: string): Promise<ScenarioRunRecord> =>
        this.scenarioGetRun(scenarioRid, runRid),
      waitForRun: (
        scenarioRid: string,
        runRid: string,
        opts?: ScenarioRunPollOptions,
      ): Promise<ScenarioRunRecord> => this.scenarioWaitForRun(scenarioRid, runRid, opts),
      applyToMain: (input: ApplyToMainInput) => this.scenarioApplyToMain(input),
    };
  }

  private async scenarioCreate(input: ScenarioCreateInput): Promise<Scenario> {
    return this.postJSON<Scenario>('/api/vertex/v1/scenarios', input);
  }

  private async scenarioApplyToMain(input: ApplyToMainInput): Promise<{ ontologyCommit: string }> {
    return this.postJSON('/api/vertex/v1/scenarios/' + encodeURIComponent(input.scenarioRid) + '/apply', {});
  }

  private async scenarioRun(
    rid: string,
    opts?: ScenarioRunOptions,
  ): Promise<ScenarioRunRecord> {
    if (opts?.streaming) {
      throw new Error('Vertex SDK: scenario-run streaming is not mounted; use polling waitForRun');
    }
    const accepted = await this.scenarioStartRun(rid);
    if (!accepted.runRid) {
      throw new Error('Vertex SDK: scenario run start response is missing runRid');
    }
    return this.scenarioWaitForRun(rid, accepted.runRid, opts);
  }

  private async scenarioStartRun(rid: string): Promise<ScenarioRunStartResponse> {
    const path = '/api/vertex/v1/scenarios/' + encodeURIComponent(rid) + '/runs';
    return this.postJSON<ScenarioRunStartResponse>(path, {});
  }

  private async scenarioGetRun(scenarioRid: string, runRid: string): Promise<ScenarioRunRecord> {
    return this.getJSON<ScenarioRunRecord>(this.scenarioRunRecordPath(scenarioRid, runRid));
  }

  private async scenarioWaitForRun(
    scenarioRid: string,
    runRid: string,
    opts: ScenarioRunPollOptions = {},
  ): Promise<ScenarioRunRecord> {
    const intervalMs = Math.max(0, opts.intervalMs ?? 1000);
    const timeoutMs = Math.max(0, opts.timeoutMs ?? 60_000);
    const deadline = Date.now() + timeoutMs;
    while (true) {
      throwIfPollingAborted(opts.signal);
      if (Date.now() >= deadline) {
        throw new VertexScenarioRunPollingTimeoutError(timeoutMs);
      }
      const run = await this.scenarioGetRunWithSignal(scenarioRid, runRid, opts.signal);
      if (isTerminalScenarioRunStatus(run.status)) {
        return run;
      }
      const remainingMs = deadline - Date.now();
      if (remainingMs <= 0) {
        throw new VertexScenarioRunPollingTimeoutError(timeoutMs);
      }
      await waitForPollDelay(Math.min(intervalMs, remainingMs), opts.signal);
    }
  }

  private async scenarioGetRunWithSignal(
    scenarioRid: string,
    runRid: string,
    signal?: AbortSignal,
  ): Promise<ScenarioRunRecord> {
    return this.getJSON<ScenarioRunRecord>(this.scenarioRunRecordPath(scenarioRid, runRid), signal);
  }

  private scenarioRunRecordPath(scenarioRid: string, runRid: string): string {
    return (
      '/api/vertex/v1/scenarios/' +
      encodeURIComponent(scenarioRid) +
      '/runs/' +
      encodeURIComponent(runRid)
    );
  }

  private async postJSON<T>(path: string, body: unknown): Promise<T> {
    const res = await this.fetchImpl(this.baseUrl + path, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(body),
    });
    const text = await res.text();
    if (!res.ok) throw new VertexHttpError(res.status, path, text);
    return JSON.parse(text) as T;
  }

  private async getJSON<T>(path: string, signal?: AbortSignal): Promise<T> {
    let res: Response;
    try {
      res = await this.fetchImpl(this.baseUrl + path, {
        method: 'GET',
        signal,
      });
    } catch (err) {
      if (signal?.aborted) {
        throw new VertexScenarioRunPollingAbortedError();
      }
      throw err;
    }
    const text = await res.text();
    if (!res.ok) throw new VertexHttpError(res.status, path, text);
    return JSON.parse(text) as T;
  }
}

function isTerminalScenarioRunStatus(status: ScenarioRunLifecycleStatus): boolean {
  return status === 'succeeded' || status === 'failed' || status === 'canceled';
}

function throwIfPollingAborted(signal?: AbortSignal): void {
  if (signal?.aborted) {
    throw new VertexScenarioRunPollingAbortedError();
  }
}

function waitForPollDelay(ms: number, signal?: AbortSignal): Promise<void> {
  throwIfPollingAborted(signal);
  if (ms <= 0) return Promise.resolve();
  return new Promise((resolve, reject) => {
    let timeoutId: ReturnType<typeof setTimeout> | null = null;
    const cleanup = () => {
      if (timeoutId !== null) {
        clearTimeout(timeoutId);
        timeoutId = null;
      }
      signal?.removeEventListener('abort', onAbort);
    };
    const onAbort = () => {
      cleanup();
      reject(new VertexScenarioRunPollingAbortedError());
    };
    timeoutId = setTimeout(() => {
      cleanup();
      resolve();
    }, ms);
    signal?.addEventListener('abort', onAbort, { once: true });
  });
}

File: sdk/go/vertex.go
// Package weavesdk is the Weave Go SDK. The vertex namespace (VTX-110)
// wires Scenario create / run / apply_to_main against the same HTTP API
// the Python and TypeScript SDKs hit. Run follows the mounted
// start-and-poll contract: POST /runs returns a run RID, then GET
// /runs/{runRid} yields the persisted lifecycle record.
//
// The transport is decoupled via the Doer interface; the unit suite
// drives the client with a httptest.Server.
package weavesdk

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// Doer is the minimal HTTP interface the client depends on so tests can
// inject httptest.Server clients.
type Doer interface {
	Do(req *http.Request) (*http.Response, error)
}

// Client is the top-level SDK entry point.
type Client struct {
	BaseURL string
	HTTP    Doer
	APIKey  string

	Vertex *VertexService
}

// New constructs a Client with sensible defaults.
func New(baseURL string, apiKey string) *Client {
	c := &Client{
		BaseURL: strings.TrimRight(baseURL, "/"),
		HTTP:    http.DefaultClient,
		APIKey:  apiKey,
	}
	c.Vertex = &VertexService{client: c, Scenarios: &ScenariosService{client: c}}
	return c
}

// VertexService is exposed at client.Vertex.
type VertexService struct {
	client    *Client
	Scenarios *ScenariosService
}

// ScenariosService is exposed at client.Vertex.Scenarios.
type ScenariosService struct {
	client *Client
}

// Scenario mirrors the wire shape returned by /api/vertex/v1/scenarios.
type Scenario struct {
	RID                  string `json:"rid"`
	CaseStudyRID         string `json:"caseStudyRid"`
	Name                 string `json:"name"`
	Status               string `json:"status"`
	Immutable            bool   `json:"immutable"`
	ParentOntologyCommit string `json:"parentOntologyCommit,omitempty"`
}

// ScenarioCreateInput names the fields used by Scenarios.Create.
type ScenarioCreateInput struct {
	CaseStudyRID         string `json:"caseStudyRid"`
	Name                 string `json:"name"`
	ParentOntologyCommit string `json:"parentOntologyCommit"`
}

// RunOptions tunes Scenarios.Run polling.
type RunOptions struct {
	PollInterval time.Duration
}

// ScenarioRunStartResponse mirrors POST
// /api/vertex/v1/scenarios/{rid}/runs.
type ScenarioRunStartResponse struct {
	RunRID string `json:"runRid"`
	Status string `json:"status"`
}

// ScenarioRunRecord mirrors the GET /api/vertex/v1/scenarios/{rid}/runs/{runRid}
// payload. Checkpoint remains a generic map so SDK callers keep all server
// details without waiting for typed retry/checkpoint helpers.
type ScenarioRunRecord struct {
	RID         string         `json:"rid"`
	ScenarioRID string         `json:"scenarioRid"`
	Status      string         `json:"status"`
	Error       string         `json:"error,omitempty"`
	Checkpoint  map[string]any `json:"checkpoint,omitempty"`
	StartedAt   time.Time      `json:"startedAt,omitempty"`
	CompletedAt *time.Time     `json:"completedAt,omitempty"`
	CreatedAt   time.Time      `json:"createdAt,omitempty"`
}

// WaitRunOptions tunes Scenarios.WaitRun.
type WaitRunOptions struct {
	PollInterval time.Duration
}

// Create POSTs to /api/vertex/v1/scenarios.
func (s *ScenariosService) Create(ctx context.Context, in ScenarioCreateInput) (*Scenario, error) {
	var out Scenario
	if err := s.client.postJSON(ctx, "/api/vertex/v1/scenarios", in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ApplyToMain POSTs to /api/vertex/v1/scenarios/{rid}/apply.
func (s *ScenariosService) ApplyToMain(ctx context.Context, scenarioRID string) (map[string]any, error) {
	out := map[string]any{}
	path := "/api/vertex/v1/scenarios/" + scenarioRID + "/apply"
	if err := s.client.postJSON(ctx, path, struct{}{}, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// Run starts a scenario run, then polls the mounted run record route until the
// run reaches a terminal status or ctx is canceled.
func (s *ScenariosService) Run(ctx context.Context, scenarioRID string, opts *RunOptions) (*ScenarioRunRecord, error) {
	accepted, err := s.StartRun(ctx, scenarioRID)
	if err != nil {
		return nil, err
	}
	if accepted.RunRID == "" {
		return nil, fmt.Errorf("vertex.scenarios.run: start response missing runRid")
	}
	waitOpts := &WaitRunOptions{}
	if opts != nil {
		waitOpts.PollInterval = opts.PollInterval
	}
	return s.WaitRun(ctx, scenarioRID, accepted.RunRID, waitOpts)
}

// StartRun POSTs to /api/vertex/v1/scenarios/{rid}/runs and returns the
// accepted run RID for polling or cancellation.
func (s *ScenariosService) StartRun(ctx context.Context, scenarioRID string) (*ScenarioRunStartResponse, error) {
	var out ScenarioRunStartResponse
	path := scenarioRunStartPath(scenarioRID)
	if err := s.client.postJSON(ctx, path, struct{}{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetRun fetches a persisted scenario-run record from the canonical
// /api/vertex/v1/scenarios/{rid}/runs/{runRid} route.
func (s *ScenariosService) GetRun(ctx context.Context, scenarioRID string, runRID string) (*ScenarioRunRecord, error) {
	var out ScenarioRunRecord
	path := scenarioRunRecordPath(scenarioRID, runRID)
	if err := s.client.getJSON(ctx, path, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// WaitRun polls GetRun until the scenario run reaches a terminal state or ctx
// is canceled. Use context.WithTimeout to bound total wait time.
func (s *ScenariosService) WaitRun(ctx context.Context, scenarioRID string, runRID string, opts *WaitRunOptions) (*ScenarioRunRecord, error) {
	interval := time.Second
	if opts != nil && opts.PollInterval > 0 {
		interval = opts.PollInterval
	}
	for {
		run, err := s.GetRun(ctx, scenarioRID, runRID)
		if err != nil {
			return nil, err
		}
		if isTerminalScenarioRunStatus(run.Status) {
			return run, nil
		}
		timer := time.NewTimer(interval)
		select {
		case <-ctx.Done():
			if !timer.Stop() {
				select {
				case <-timer.C:
				default:
				}
			}
			return nil, ctx.Err()
		case <-timer.C:
		}
	}
}

func scenarioRunRecordPath(scenarioRID string, runRID string) string {
	return "/api/vertex/v1/scenarios/" + url.PathEscape(scenarioRID) + "/runs/" + url.PathEscape(runRID)
}

func scenarioRunStartPath(scenarioRID string) string {
	return "/api/vertex/v1/scenarios/" + url.PathEscape(scenarioRID) + "/runs"
}

func isTerminalScenarioRunStatus(status string) bool {
	return status == "succeeded" || status == "failed" || status == "canceled"
}

func (c *Client) postJSON(ctx context.Context, path string, body any, out any) error {
	buf, err := json.Marshal(body)
	if err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.BaseURL+path, bytes.NewReader(buf))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	c.applyAuth(req)
	resp, err := c.HTTP.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode/100 != 2 {
		return fmt.Errorf("vertex SDK: %d %s", resp.StatusCode, string(respBody))
	}
	if out == nil || len(respBody) == 0 {
		return nil
	}
	return json.Unmarshal(respBody, out)
}

func (c *Client) getJSON(ctx context.Context, path string, out any) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.BaseURL+path, http.NoBody)
	if err != nil {
		return err
	}
	c.applyAuth(req)
	resp, err := c.HTTP.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode/100 != 2 {
		return fmt.Errorf("vertex SDK: %d %s", resp.StatusCode, string(respBody))
	}
	if out == nil || len(respBody) == 0 {
		return nil
	}
	return json.Unmarshal(respBody, out)
}

func (c *Client) applyAuth(req *http.Request) {
	if c.APIKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.APIKey)
	}
}
